pkg/iam.go

```diff
@@ -0,0 +1,154 @@
+package pkg
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/servicequotas"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type IAMClient interface {
+	ListRolesPagesWithContext(ctx aws.Context, input *iam.ListRolesInput, fn func(*iam.ListRolesOutput, bool) bool, opts ...request.Option) error
+}
+
+type ServiceQuotasClient interface {
+	GetServiceQuotaWithContext(ctx aws.Context, input *servicequotas.GetServiceQuotaInput, opts ...request.Option) (*servicequotas.GetServiceQuotaOutput, error)
+}
+
+type AWSIAMClient struct {
+	iam *iam.IAM
+}
+
+func (c *AWSIAMClient) ListRolesPagesWithContext(ctx aws.Context, input *iam.ListRolesInput, fn func(*iam.ListRolesOutput, bool) bool, opts ...request.Option) error {
+	return c.iam.ListRolesPagesWithContext(ctx, input, fn, opts...)
+}
+
+type AWSServiceQuotasClient struct {
+	sq *servicequotas.ServiceQuotas
+}
+
+func (c *AWSServiceQuotasClient) GetServiceQuotaWithContext(ctx aws.Context, input *servicequotas.GetServiceQuotaInput, opts ...request.Option) (*servicequotas.GetServiceQuotaOutput, error) {
+	return c.sq.GetServiceQuotaWithContext(ctx, input, opts...)
+}
+
+var (
+	// Prometheus metrics
+	IamRolesUsed = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "iam", "roles_used"),
+		"Number of IAM roles used in the account.",
+		[]string{"aws_account_id"}, nil,
+	)
+	IamRolesQuota = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "iam", "roles_quota"),
+		"IAM role quota for the account.",
+		[]string{"aws_account_id"}, nil,
+	)
+	IamRolesUsagePercent = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "iam", "roles_usage_percent"),
+		"Percentage of IAM roles used relative to the quota.",
+		[]string{"aws_account_id"}, nil,
+	)
```

Comment:
do we need to export this computed value since we can get it via `IamRolesUsed` and `IamRolesQuota`?

Reply:
Good question. While it’s true the usage percent could be computed from the other two metrics in Prometheus.

Exposing it directly helps simplify alerting rules. It also aligns with how other exporters expose `*_usage_percent metrics` for easier consumption for the IAM role capacity monitoring we are going to integrate with Grafana.

Reply:
ok, removed that metric